### PR TITLE
feat: support using rootUri for workspaceFolders and handle workspaceChangeEvent

### PR DESCRIPTION
**Pull request by shruti0085**
## Description
Currently in a lot of downstream servers, the `workspace.getWorkspaceFolder` method is used for path resolution. This only relied on the `initializeParams.workspaceFolder` value which did not take into account workspace folder changes(did not listen to `didChangeWorkspaceFolders`) that could have happened during the session.
In addition with VS client as mentioned in https://github.com/aws/language-servers/pull/1210, the workspaceFolder value is not set(more details in that PR) however it does specify that information via `rootUri` and `rootPath`. The utility class from that PR has now been moved centrally to this repo.
A new `getAllWorkspaceFolders()` call has been added as a common utility that servers downstream can use as single source of truth to get latest information of workspaceFolders available to the lsp server.
A follow-up PR to language-servers repo will utilize this change.

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: runtimes/server-interface/workspace.ts [hunk 1]
```diff
@@ -19,6 +19,7 @@ export type Workspace = {
     getTextDocument: (uri: string) => Promise<TextDocument | undefined>
     getAllTextDocuments: () => Promise<TextDocument[]>
     getWorkspaceFolder: (uri: string) => WorkspaceFolder | null | undefined
+    getAllWorkspaceFolders: () => WorkspaceFolder[]
```
Comment (justinmk3) [inline runtimes/server-interface/workspace.ts:22]: just make it plural. helps discoverability because it will appear next to the singular case.
```suggestion
    getWorkspaceFolders: () => WorkspaceFolder[]
```

Comment (justinmk3) [inline runtimes/server-interface/workspace.ts:22]: just make it plural. helps discoverability because it will appear next to the singular case. docstrings could be used for both of these cases, meanwhile...


```suggestion
    /** Gets the current(?) workspace folder as decided by ... */
    getWorkspaceFolder: (uri: string) => WorkspaceFolder | null | undefined
    /** Gets all workspace folders known to <something> */
    getWorkspaceFolders: () => WorkspaceFolder[]
```

Comment (shruti0085) [reply runtimes/server-interface/workspace.ts:22]: Yes I followed the existing pattern. I will add the doc strings for clarity

File: runtimes/runtimes/base-runtime.ts [hunk 2]
```diff
@@ -124,8 +124,8 @@ export const baseRuntime = (connections: { reader: MessageReader; writer: Messag
     const workspace: Workspace = {
         getTextDocument: async uri => documents.get(uri),
         getAllTextDocuments: async () => documents.all(),
-        getWorkspaceFolder: _uri =>
-            lspRouter.clientInitializeParams!.workspaceFolders && lspRouter.clientInitializeParams!.workspaceFolders[0],
+        getWorkspaceFolder: _uri => lspRouter.getAllWorkspaceFolders() && lspRouter.getAllWorkspaceFolders()[0],
+        getAllWorkspaceFolders: () => lspRouter.getAllWorkspaceFolders(),
```
Comment (ege0zcan) [inline runtimes/runtimes/base-runtime.ts:128]: I think the addition of `getAllWorkspaceFolders` makes this PR more of a feature rather than a fix

I am also noticing that none of the `Workspace` functionality is documented in the  readme, maybe we can start to fix things by documenting `getWorkspaceFolder` and `getAllWorkspaceFolders` as part of this PR or as a followup

File: runtimes/runtimes/lsp/router/initializeUtils.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+            return [{ name: folderName, uri: params.rootUri }]
+        }
+        if (params.rootPath) {
+            const parsedUri = URI.parse(params.rootPath)
+            const folderName = getFolderName(parsedUri)
+            return [{ name: folderName, uri: parsedUri.toString() }]
+        }
+        return []
+    } catch (error) {
+        console.log(`Error occurred when determining workspace folders: ${error}`)
```
Comment (ege0zcan) [inline runtimes/runtimes/lsp/router/initializeUtils.ts:29]: potentially we could use `console.warn` or `console.error` here

**Review by ege0zcan — APPROVED** [review]

File: runtimes/runtimes/lsp/router/initializeUtils.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
+import { InitializeParams, WorkspaceFolder } from 'vscode-languageserver-protocol'
+import path from 'path'
+import { URI } from 'vscode-uri'
+import { RemoteConsole } from 'vscode-languageserver'
+
+export function getWorkspaceFoldersFromInit(console: RemoteConsole, params?: InitializeParams): WorkspaceFolder[] {
```
[the remark this answers is not in the document]
Comment (shruti0085) [reply runtimes/runtimes/lsp/router/initializeUtils.ts:6]: That's fair. I added it because of the edge cases as I was writing the tests. Currently from the location where this is invoked it should never be null but considering this is a utility call shared, I am inclined to leave it as it.

**Review by awschristou — APPROVED** [review]
